runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -1224,20 +1711,125 @@ TR::Register *J9::Power::TreeEvaluator::awrtbariEvaluator(TR::Node *node, TR::Co
+   if (cg->comp()->getOption(TR_EnableFieldWatch))
+      {
+      TR::TreeEvaluator::rdWrtbarHelperForFieldWatch(node, cg, sideEffectRegister, valueReg);
+      }
+   // The Value Node, or the second child is not decremented here. The store evaluator also uses it, and decrements it.
+   cg->decReferenceCount(sideEffectNode);
+   return TR::TreeEvaluator::dstoreEvaluator(node, cg);
+   }
+
+TR::Register *J9::Power::TreeEvaluator::dwrtbariEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   /* rdbar & wrtbar nodes contain the sideEffect Child Node. 
+      During Field Watch we evaluate the sideEffect and it's used within the process.
+      The evaluation of the remaining children, as well as the load/store operation occur within the respective evaluator. */
+   
+   TR::Node *sideEffectNode = node->getThirdChild();
+   TR::Register *valueReg = cg->evaluate(node->getSecondChild());
+   TR::Register *sideEffectRegister = cg->evaluate(node->getThirdChild());
+   if (cg->comp()->getOption(TR_EnableFieldWatch))
+      {
+      TR::TreeEvaluator::rdWrtbarHelperForFieldWatch(node, cg, sideEffectRegister, valueReg);
+      }
+   // The Value Node, or the second child is not decremented here. The store evaluator also uses it, and decrements it.
+   cg->decReferenceCount(sideEffectNode);
+   return TR::TreeEvaluator::dstoreEvaluator(node, cg);
+   }
+
 TR::Register *J9::Power::TreeEvaluator::irdbarEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   return TR::TreeEvaluator::irdbariEvaluator(node, cg);
+   /* rdbar & wrtbar nodes contain the sideEffect Child Node. 
+      During Field Watch we evaluate the sideEffect and it's used within the process.
+      The evaluation of the remaining children, as well as the load/store operation occur within the respective evaluator. */
+
+   TR::Node *sideEffectNode = node->getFirstChild();
+   TR::Register * sideEffectRegister = cg->evaluate(sideEffectNode);
+   if (cg->comp()->getOption(TR_EnableFieldWatch))
+      {
+      TR::TreeEvaluator::rdWrtbarHelperForFieldWatch(node, cg, sideEffectRegister, NULL);
+      }
+   cg->decReferenceCount(sideEffectNode);
+   return TR::TreeEvaluator::iloadEvaluator(node, cg);
    }
 
 TR::Register *J9::Power::TreeEvaluator::irdbariEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   TR_ASSERT(TR::Compiler->om.readBarrierType() != gc_modron_readbar_none, "iReadBarrierEvaluator");
-#ifdef OMR_GC_CONCURRENT_SCAVENGER
-   TR_ASSERT(cg->comp()->useCompressedPointers(), "irdbarEvaluator is expecting compressed references");
+   /* rdbar & wrtbar nodes contain the sideEffect Child Node. 
+      During Field Watch we evaluate the sideEffect and it's used within the process.
+      The evaluation of the remaining children, as well as the load/store operation occur within the respective evaluator. */   
+   TR::Node *sideEffectNode = node->getFirstChild();
+   TR::Register *sideEffectRegister = cg->evaluate(sideEffectNode);
```

Comment:
Who is going to decrement this child?

Reply:
Either the iloadEvaluator or concurrent scavenger sequence which occurs right after reporting.

Reply:
Either iloadEvaluator, or the concurrent scavenger sequence.